### DISCUSSION TITLE
How to restore a saved dmp file using MQTT

### DISCUSSION THREAD

**Discussion by carefulcomputer** [arendst/Tasmota#20332] [Q&A]
I am writing an automation in node-red which downloads all tasmota configs and restores them in case something goes wrong. I have the automation to download the configs binary (dmp file) from all tasmota using /dl endpoint. I verified that dmp files work by manually uploading the file to tasmota through web interface. 
However, when I am trying to restore using /u2?fsz= endpoint (file upload using http post), i see that web page returns 'successful' response but the settings themeselves are not applied. What am I doing wrong ? If webupload is not possible, is there a MQTT command I send with the binary dmp file for it to restore the settings ? I can't seem to find any documentation about restoring config programmatically in tasmota commands page https://tasmota.github.io/docs/Commands/ .

**Comment by sfromis** [comment]
The [WebGetConfig](https://tasmota.github.io/docs/Commands/#webgetconfig) command exists as an optional feature.

Anyway, with the tip on Discord, you may have an avenue already.

**Comment by carefulcomputer** [comment]
Thanks for response @sfromis . WebGetCofig seems like a way to 'pull' config from a web url. I am trying to 'push' from my server using HTTP/MQTT. I tried looking at the decode-config, but I couldn't find what else it is doing to succeed where my request is failing. :( . I copied all headers and uploading the file as such

```
let form = new FormData();
form.append('u2', msg.payload); // msg.payload is a buffer containing dmp file content read in previous file node

const result = await axios.post(`${msg.url}u2?fsz=${msg.payload.length}`, form, {
    headers: {
        'Authorization': 'Basic BASEENCODED_USER_PASS',
        'Referer': 'http://192.168.2.15/rs?',
        'Origin': 'http://192.168.2.15',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8',
        'Accept-Language': 'en-US,en',
        'Sec-Gpc' : 1
    }});

msg.result = 'Failed';
if (result.status === 200 && result.data.includes('Successful')) {
    msg.result = 'Success';
}
return msg;
```

When I push file using axios (in node-red), I can see the response HTML has 'success' message but the config is still not updating. I must be missing something. Just don't know what. I hope this was documented somewhere more clearly.

**Reply by barbudor** [reply]
AFAIK the Basic Auth method is not supported
If you have a password set, you must insert the user (admin) and password in the url like in the documentation


Just to give you some ideas, here is what I have chosen:
I have a nightly script on my server that pull the configuration for every device and save that in a local git repo. Latest conf files are available through http.
All my devices rune a custom binary which include wifi, defaults mqtt credentials and a Rule1  which on boot use webconfig to retrieve the last known configuration for that device (%id.dmp)
If any thing goes wrong, I can send a 'reset 1' or use any of the recovery mechanism which will revert to the hard-coded configuration which in turn will reload the last good known configuration
Full self re-provisinning

**Reply by sfromis** [reply]
I posted the command for context around your statement:
> I can't seem to find any documentation about restoring config programmatically in tasmota commands page

Looks like you did not try the tip I gave you on Discord, to use the method `decode-config` is using, specifically at the point of the comment I mentioned.

```
    # get restore config page first to set internal Tasmota vars
    responsecode, body = get_tasmotaconfig('rs?', http_host, http_port, http_username, http_password, contenttype='text/html')
```

**Reply by carefulcomputer** [reply]
@sfromis, I am not very good with python, but from what I could understand, that method is grabbing referrer etc. I did same.
@barbudor, the headers I am sending are literally copied from the chrome web inspector.

**Reply by sfromis** [reply]
The word "referer" was not even part of what I posted. It does not setup a referer, it makes a http request.

**Reply by carefulcomputer** [reply]
i am not very good with python, can you please share exactly what the script is doing ? when I track down that function i see this

url = make_url(host, port, cmnd)
referer = make_url(host, port)
auth = (username, password)

from my limited knowledge of python, i can't find out what is different in my post request vs what the code is doing ?  do you see something I missed ?

**Reply by sfromis** [reply]
Well, I'm not good with Python either, but I already told you what happens. It is making an additional http request, as clearly mentioned by the comment line. It is **not** about replicating stuff in headers.

**Comment by arendst** [comment]
Have a look at folder tools/mqtt-file. Here you find plain mqtt examples of saving and restoring .dmp files.

**Reply by carefulcomputer** [reply]
Thanks @arendst . That looks like the exact thing I need. It will be awesome if full protocol is updated in tasmota docs.  Also another question, I tried to replicate the browser file uploading by trying to copy exact headers and uploading the file to /u2. I even get a 'successful' message after the file upload, but tasmota doesn't use the settings. Chrome inspector doesn't show the full payload,  but I gathered (maybe incorrectly) that only item submitted is a variable named 'u2' with the value as the file itself. Is there any other parameter I need to submit at http endpoint ?
